File: BackgroundThread/MyBackgroundThread/java/app/src/main/java/com/dicoding/picodiploma/mybackgroundthread/MainActivity.java
package com.dicoding.picodiploma.mybackgroundthread;

import android.os.AsyncTask;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.TextView;

import java.lang.ref.WeakReference;
import java.util.concurrent.Callable;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {

    private TextView tvStatus;
    private TextView tvDesc;

    private final static String INPUT_STRING = "Halo Ini Demo AsyncTask!!";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvStatus = findViewById(R.id.tv_status);
        tvDesc = findViewById(R.id.tv_desc);

        new Thread(() -> {
            tvStatus.setText(R.string.status_pre);
            tvDesc.setText(INPUT_STRING);

            String output = INPUT_STRING + " Selamat Belajar!!";
            Log.d("LOG_ASYNC", "status : doInBackground");

            try {
                Thread.sleep(2000);
                runOnUiThread(() -> {
                    tvStatus.setText(R.string.status_post);
                    tvDesc.setText(output);
                });

            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }).start();
    }
}

File: BackgroundThread/MyBackgroundThread/kotlin/app/src/main/java/com/dicoding/picodiploma/mybackgroundthread/MainActivity.kt
package com.dicoding.picodiploma.mybackgroundthread

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val INPUT_STRING = "Halo Ini Demo AsyncTask!!"
        private const val LOG_ASYNC = "DemoAsync"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tv_status.setText(R.string.status_pre)
        tv_desc.text = INPUT_STRING

        GlobalScope.launch(Dispatchers.Default) {
            //background thread
            val input = INPUT_STRING

            var output: String? = null

            Log.d(LOG_ASYNC, "status : doInBackground")
            try {
                // Input stringnya ditambahkan dengan string ' Selamat Belajar!!"
                output = "$input Selamat Belajar!!"

                /*
                Sleep thread digunakan untuk simulasi bahwa ada proses yang sedang berjalan selama 2 detik
                2000 miliseconds = 2 detik
                */
                delay(2000)

                //pindah ke Main thread untuk update UI
                withContext(Dispatchers.Main) {
                    tv_status.setText(R.string.status_post)
                    tv_desc.text = output
                }

            } catch (e: Exception) {
                Log.d(LOG_ASYNC, e.message.toString())
            }
        }
    }
}
